web/hooks/push.go

```diff
@@ -0,0 +1,259 @@
+	}
+	return assignments
+}
+
+func extractChanges(changes []string, modifiedAssignments map[string]bool) {
+	for _, changedFile := range changes {
+		index := strings.Index(changedFile, "/")
+		if index == -1 {
+			// ignore root-level files
+			continue
+		}
+		// we assume the first path component holds the assignment name
+		name := changedFile[:index]
+		if name == "" {
+			// ignore names that start with "/" or empty names
+			continue
+		}
+		modifiedAssignments[name] = true
+	}
+}
+
+// runAssignmentTests runs the tests for the given assignment pushed to repo.
+func (wh GitHubWebHook) runAssignmentTests(assignment *pb.Assignment, repo *pb.Repository, course *pb.Course, payload *github.PushEvent) *score.Results {
+	runData := &ci.RunData{
+		Course:     course,
+		Assignment: assignment,
+		Repo:       repo,
+		// TODO(Espeland): If a student for some reason pushes to a remote branch with a different name than
+		// the local one, this will fail.
+		// E.g. if a student pushes to remote branch "branch1", but their code is on the local branch "branch2",
+		// QF will try to checkout the wrong local branch.
+		// The payload contains no information on the local branch, and it therefore seems like there is no good workaround.
+		// We must make sure that if this scenario happens, QF can handle it.
+		BranchName: getBranchName(payload.GetRef()),
+		CommitID:   payload.GetHeadCommit().GetID(),
+		JobOwner:   payload.GetSender().GetLogin(),
+	}
+	if assignment.GradedManually() {
+		wh.logger.Debugf("Assignment %s for course %s is manually reviewed", assignment.Name, course.Name)
+		if _, err := runData.RecordResults(wh.logger, wh.db, nil); err != nil {
+			wh.logger.Error(err)
+		}
+		return nil
+	}
+	ctx, cancel := assignment.WithTimeout(ci.DefaultContainerTimeout)
+	defer cancel()
+	results, err := runData.RunTests(ctx, wh.logger, wh.runner)
+	if err != nil {
+		wh.logger.Errorf("Failed to run tests for assignment %s for course %s: %v", assignment.Name, course.Name, err)
+	}
+	wh.logger.Debug("ci.RunTests", zap.Any("Results", log.IndentJson(results)))
+	if _, err = runData.RecordResults(wh.logger, wh.db, results); err != nil {
+		wh.logger.Error(err)
+	}
+	return results
+}
+
+// updateLastActivityDate sets a current date as a last activity date of the student
+// on each new push to the student repository.
+func (wh GitHubWebHook) updateLastActivityDate(userID, courseID uint64) {
+	query := &pb.Enrollment{
+		UserID:           userID,
+		CourseID:         courseID,
+		LastActivityDate: time.Now().Format("02 Jan"),
+	}
+
+	if err := wh.db.UpdateEnrollment(query); err != nil {
+		wh.logger.Errorf("Failed to update the last activity date for user %d: %v", userID, err)
+	}
+}
+
+// getBranchName gets the branch name from a push event
+func getBranchName(ref string) string {
+	components := strings.Split(ref, "/")
+	return components[len(components)-1]
+}
+
+// isDefaultBranch returns true if a push event came from repository's default branch.
```

Comment:
Revise comment: returns true if a push event is for a repository's default branch.